src/polyfill/bigint64array.js

```diff
@@ -0,0 +1,91 @@
+            if(!ArrayBuffer.isView(array)){
+                array = new Uint32Array(array);
+            }
+            let proxy = new Proxy({
+                slice: function(min, max) {
+                    var new_buf = array.slice(min * 2, max *2);
+                    return createBigInt64Array(new_buf);
+                },
+                subarray: function(min, max) {
+                    var new_buf = array.subarray(min * 2, max *2);
+                    return createBigInt64Array(new_buf);
+                },
+                [Symbol.iterator]: function*() {
+                    for (var i = 0; i < (array.length)/2; i++) {
+                        yield partsToBigInt(array[2*i], array[2*i+1]);
+                    }
+                },
+                buffer : array.buffer,
+                byteLength : array.byteLength,
+                offset : array.byteOffset / 2,
+                copyWithin: function(target, start, end) {
+                    array.copyWithin(target*2, start * 2, end*2);
+                    return proxy;
+                },
+                set: function(source, targetOffset) {
+                    if(2*(source.length + targetOffset) > array.length) {
+                        // This is the Chrome error message
+                        // Firefox: "invalid or out-of-range index"
+                        throw new RangeError("offset is out of bounds");
+                    }
+                    for (var i = 0; i < array.length; i++) {
+                        var value = source[i];
+                        var pair = bigIntToParts(BigInt(value));
+                        array.set(pair, 2*(targetOffset + i));
+                    }
+                }
+                }, {
+                get: function(target, idx, receiver) {
+                    if (typeof idx !== "string" || !/^\d+$/.test(idx)) {
+                        return Reflect.get(target, idx, receiver);
+                    }
+                    var lower = array[idx * 2];
+                    var upper = array[idx * 2 + 1];
+                    return partsToBigInt(lower, upper);
+                },
+                set: function(target, idx, value, receiver) {
+                    if (typeof idx !== "string" || !/^\d+$/.test(idx)) {
+                        return Reflect.set(target, idx, value, receiver);
+                    }
+                    if (typeof value !== "bigint") {
+                        // Chrome error message, Firefox has no "a" in front if "BigInt".
+                        throw new TypeError(`Cannot convert ${value} to a BigInt`);
+                    }
+                    var pair = bigIntToParts(value);
+                    array.set(pair, 2*idx);
+                }
+            });
+            return proxy;
+        }
+        return createBigInt64Array;
+    }
+
+    var BigUint64Array = createBigIntArrayShim(partsToBigIntUnsigned);
+    var BigInt64Array = createBigIntArrayShim(partsToBigIntSigned);
```

Comment:
These vars will cause `BigInt64Array` to always be undefined in the check higher up, because the `var` has global scope`. As a result I think this code path will always end up being taken, which means if we include the polyfill code we'd never get the fast code path.

Reply:
I don't understand `var`. What's the right way to deal with the scoping? Maybe:
```
var BigUint64ArrayShimmed;
var BigInt64ArrayShimmed;

if(BigUint64Array !== "undefined"){
	BigUint64ArrayShimmed = BigUint64Array;
	BigInt64ArrayShimmed = BigInt64Array;
} else {
	#include polyfill/bigint64array.js
}
HEAP64 = new BigInt64ArrayShimmed(buf);
HEAPU64 = new BigUint64ArrayShimmed(buf);
```

Reply:
Can you just drop the `var` completely?